js/fetch.ts

```diff
@@ -20,15 +20,72 @@ import {
 } from "./fetch_types";
 import { TextDecoder } from "./text_encoding";
 
+type HeadersInit = Array<[string, string]> | { [K: string]: string };
+
+interface Header {
+  name: string;
+  value: string;
+}
+
 class DenoHeaders implements Headers {
+  private readonly headerList: Header[] = [];
+
+  constructor(init?: HeadersInit) {
+    if (init) {
+      this._fill(init);
+    }
+  }
+
+  private _append(header: Header): void {
+    // TODO(qti3e) Check header based on the fetch spec.
+    this._appendToHeaderList(header);
+  }
+
+  private _appendToHeaderList(header: Header): void {
+    const lowerCaseName = header.name.toLowerCase();
+    for (let i = 0; i < this.headerList.length; ++i) {
+      if (this.headerList[i].name.toLowerCase() === lowerCaseName) {
+        header.name = this.headerList[i].name;
+      }
+    }
+    this.headerList.push(header);
+  }
+
+  private _fill(init: HeadersInit): void {
+    if (Array.isArray(init)) {
+      for (let i = 0; i < init.length; ++i) {
+        const header = init[i];
+        if (header.length !== 2) {
+          throw new TypeError();
+        }
```

Comment:
Necessary? Multiple values with the same key are allowed in http headers.

Reply:
@ry This implementation already supports multiple values for the same key, `header.length !== 2)` is based on this https://fetch.spec.whatwg.org/#concept-headers-fill

Reply:
Oh ok - then can you add a test where it shows this throwing a TypeError?